### ISSUE TITLE
Event sequencer command

### ISSUE THREAD

**Issue by egaltier** [pcdshub/Bug-Reports-and-Requests#21]
## Feature Request
Python command to trigger a single shot from the FEL from the event sequencer.

## Context
For example, when doing imprint, we need to raster the target, change the attenuation and take a single shot. Usually, taking a single shot means setting the pulse picker to flip/flop mode (via mec_pulsepicker.flipflop()) and then ask the event sequencer to trigger the pulse picker via the event code 168. We can do it manually via the event sequencer GUI, but for scripting purposes would need the python command for it.

## Hutch
MEC

**Comment by silkenelson** [comment]
Do you need a single shot or do you want an 'imprint' plan/macro? CXI will also need one (update of their currently existing, but possibly broken imprint scripts), we should figure out what this function/plan should do and make it work for both hutches.

**Comment by egaltier** [comment]
In general, we do single shot for various purposes, not just imprint. This being said, we should definitely have  single script that works for everybody for the imprints. But in addition to that, we should have the capability to 'call' for a single shot. For example, we usually have rastering script coupled to the DAQ that move to target, take a synchronized xray+optical laser shot, then move to next target, and repeat. This single shot concept is definitely going to be used in this kind of situation. We had already a bluesky script that Tyler made to do this whole sequence so maybe that the right function to call for the single shot can be taken from this script?

**Comment by ZLLentz** [comment]
This should be pretty straighforward, all the knobs are there. Something like:
```
def single_shot(pp, seq):
    seq.sequence.put_sequence([[168, 1, 0, 0]])
    pp.flipflop(wait=True)
    seq.trigger()
```

**Comment by egaltier** [comment]
`pp` could be `mec_pulsepicker` as defined by Tyler but what would be the equivalent for `seq`? Could this clean solution be used when the sequence becomes more complicated, for example when adding the laser triggers, the slow cameras triggers and other auxiliary diagnostics trigger (like gas jet) to be repeated for few 100's of times? I guess is all comes to defining the sequence it self like we do with the GUI? @slactjohnson knows this problem very well :p

**Comment by slactjohnson** [comment]
Yes, pp is the alias I made for the MEC pulse picker. In MEC the equivalent for `seq` is, well, `seq`. :) This is already available. We should talk more about doing single shots, since I think that the current laser BlueSky plans already cover this use case. I think we even have a laser raster script, though I don't think I got the opportunity to test it much. However, perhaps we should have a separate plan for something like this since it's a little weird to take an x-ray only shot with a laser script.

**Comment by silkenelson** [comment]
@slactjohnson Do you think this can be closed? If there is any work left to do in met-python, feel free to pull Nick into this unless you have higher priority work for him.

**Comment by slactjohnson** [comment]
@silkenelson Yes, I believe that this can be closed, but I'll wait for confirmation from @egaltier

**Comment by egaltier** [comment]
I don't know if this is solved 'per se' (meaning this is included in the main python release). I wrote a command to switch to a 5Hz Pulse Picker that looks like this: 
```
alignment_mode = [[168, 22, 0, 0],[169,  2, 0, 0]]
def pulse_picker(rate=5):
        # reset the mode for the pp
        pp.put(0)
        # set the flip/flop mode
        pp.put(2)
        # push the sequence
        seq.sequence.put_seq(alignment_mode)
        seq.play_mode.put(2)
        seq.start()
```
It is fixed to a 5Hz rep rate which is not perfect (but on the other side, our max rep rated laser is 5Hz so it works for us). Having this more versatile to any rep rate (adjusting the sequence delta as a function of the rate I guess) and pushing this into the main python code would make me think that this thread is solved since I am not aware of any other command that is doing a similar thing. To switch to 120 Hz then we push `x.start_seq(120)` which is available in the main python, but this just leave the pulse picker fully open.
